### PR TITLE
TINY-9268: Extend menu button and toolbar button APIs to update the menu button icon and toolbar button icon respectively

### PR DESCRIPTION
**Pull request by hamza0867**
Related Ticket: TINY-9268

Description of Changes:
* Extend menu button and toolbar button APIs to update the menu button icon and toolbar button icon respectively

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

[TINY-9268]: https://ephocks.atlassian.net/browse/TINY-9268?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts:165]: Should this be `menubutton-display-events` I think the name here is used for tooling.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:66]: One here is a block the other one is not should the setText and setIcon be the same both just lambdas without block since they have just one single statement in them?

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:160]: Remove the old code here.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:138]: Maybe add a comment why this is required since we want to keep the width when `setText` is executed but that might not be obvious if you read this out of context. Or maybe describe it with code since this is used in a few places maybe it's better to have say a function that does `ToolbarUtils.forceInitialSize(comp)` or similar that does the code and also describes why it's there.

```suggestion
            Css.set(comp.element, 'width', Css.get(comp.element, 'width'));
```

File: modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts:167]:
```suggestion
            Css.set(comp.element, 'width', Css.get(comp.element, 'width'));
```

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:160]: I am waiting for @ephox-mogran to review this code and see if this approach is the right approach or not. That is why I am keeping these comments until I get a review from him.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts]: Agreed. Done!

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts]: Done. I used UiUtils since it is not used only in toolbar but also in dropdown.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (ephox-mogran) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:306]: Try to exist using direct child relationships for identifying descendants - it makes everything too coupled to the existing DOM structure. Is `comp.components()[0]` the same as the `descendant('span')` above? Could a similar approach be used instead?

File: modules/tinymce/src/plugins/autosave/main/ts/ui/Buttons.ts
Comment (ephox-mogran) [inline modules/tinymce/src/plugins/autosave/main/ts/ui/Buttons.ts:6]: This change seems unexpected. `Menu.MenuItemInstanceApi` sounds like the API of a single menu item, whereas this is in Buttons. Is this really intended?

**Review by ephox-mogran — COMMENTED** [review]
Seems like it's almost there, but there are a few questions I have. We might try and organise a session to go through them.

File: modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts
Comment (ephox-mogran) [reply modules/tinymce/src/themes/silver/main/ts/ui/dropdown/CommonDropdown.ts:166]: This also looks like a duplicate name with ToolbarButtons

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (ephox-mogran) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:157]: I think we can see two options here 

* if we still need to respond to the channel `r` broadcasts, then instead of being reflecting, this should be receiving, and it should respond to the message by emitting the custom events: `updateMenuText` and `updateMenuIcon` with the data passed as part of the message
* if we don't need to respond to the channel `r` broadcasts, then we should just remove this reflecting altogether.

Either way, what we have here isn't likely to be correct. If it receives a broadcast, it's going to reset the text and icon back to the original, and clobber any of the subsequent replacing calls, based on no information (it isn't using the information from the reflecting message)

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (ephox-mogran) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:216]: Previously, these were the behaviours defined in buttonBehaviours
 
 * disabling 
 * receiving for read-only (if we switch to using receiving for updating menu text etc., we'll need to combine these configs)
 * mousedown handler for focusing the button
 * reflecting configuration if required 
 
 But this code path wasn't using them. It is now. So in addition to the new events you've defined, these behaviours will be added to the button. You need to ensure that that change is intentional. Was it a bug before that readonly wasn't respected, or that disabled wasn't expected etc. 
 
 Also, try to track down what is passing `receiver`, and see if the receiving / reflecting configs are still necessary.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts]: Agreed. Changed to use SelectorFind.descendant

**Review by vpyshnenko — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (TheSpyder) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:211]: This is a red flag for me - behaviours should always go inside the `derive` call. Not doing that means there's probably a bug lurking somewhere.

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/toolbar/ToolbarDrawerToggleTest.ts
Comment (TheSpyder) [inline modules/tinymce/src/themes/silver/test/ts/browser/editor/toolbar/ToolbarDrawerToggleTest.ts:41]: What's going on here? Why do we suddenly need to know the difference between Chrome and Edge?

File: modules/tinymce/src/themes/silver/test/ts/module/ToolbarUtils.ts
Comment (TheSpyder) [inline modules/tinymce/src/themes/silver/test/ts/module/ToolbarUtils.ts:33]: This doesn't seem related to the feature you added 🤔

**Review by TheSpyder — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/themes/silver/test/ts/browser/editor/toolbar/ToolbarDrawerToggleTest.ts
Comment (spocke) [reply modules/tinymce/src/themes/silver/test/ts/browser/editor/toolbar/ToolbarDrawerToggleTest.ts:41]: We discussed a weird flake that is only happening on Edge. So trying to skip it for now it's unrelated to the test. Not sure what to do with these flakes it takes lots of time to investigate them. I would probably not be something a user will run into.

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:299]: I wonder if this could just be emitted on the `comp` instead of the `buttonComp` the event broadcast should be sent down to all the descendants. Or do you think there is a risk that something on the actual comp would also listen to the same broadcasts?

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:299]: Unfortunately that doesn't work, I know it because a test case fails, but I don't know exactly why it doesn't work.

File: modules/tinymce/src/themes/silver/test/ts/module/ToolbarUtils.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/test/ts/module/ToolbarUtils.ts]: Indeed it seems I needed to make this change to pass some tests locally and forgot to clean it up. I just cleaned it up in a new commit

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (TheSpyder) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:211]: That wasn't clear - please add a comment about why we're doing it this way, because (as far as I can tell) it's different to every other behaviour specification we have.

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts
Comment (hamza0867) [reply modules/tinymce/src/themes/silver/main/ts/ui/toolbar/button/ToolbarButtons.ts:214]: Done!

**Comment by TheSpyder** [comment]
I'm pretty sure the failing test is not caused by this PR, develop is passing so let's see what happens

**Pull request merged**
